### PR TITLE
Fix incremental build for single-file

### LINKED ISSUE #12871: Building a self contained, then single file version of app will build bad single file version.
In working on a blog post for single file, I have been building a set of different publish versions of the app. I had been deleting the bin and obj folders between each trim mode, but not between self contained and single file. The latter failed if the bin & obj had not been cleaned first.
It seems like dotnet publish should include a clean as part of its build steps - if it can't handle what is already there, then it needs to do a clean. It should not fail if the existing state is not what it expected.

Failure occurs on windows and Linux (in WSL 2).

```
rm -f -r -d bin
rm -f -r -d obj
dotnet publish -r linux-x64 -p:PublishTrimmed=True -c Release -p:TrimMode=CopyUsed   -o ../AssemblyTrimmed
# rm -f -r -d bin
# rm -f -r -d obj
dotnet publish -r linux-x64 -p:PublishTrimmed=True -c Release -p:TrimMode=CopyUsed   -p:PublishSingleFile=true  -o ../AssemblyTrimmedSF  
```
The latter will succeed in publishing but fail at runtime with 
```
A fatal error was encountered. The library 'libhostpolicy.so' required to execute the application was not found in '/usr/share/dotnet'.
Failed to run as a self-contained app.
  - The application was run as a self-contained app because '/home/sam/FullTrimSF/hello_test.runtimeconfig.json' was not found.
  - If this should be a framework-dependent app, add the '/home/sam/FullTrimSF/hello_test.runtimeconfig.json' file and specify the appropriate framework.
```
if the second set of rm are commented out.

### PR DESCRIPTION
**Pull request by agocke**
Fixes #12871

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets [hunk 1]
```diff
@@ -625,7 +626,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     ============================================================
     -->
   <Target Name="_ComputeNETCoreBuildOutputFiles"
-          DependsOnTargets="_GetAppHostPaths;_GetComHostPaths;_GetIjwHostPaths"
+          DependsOnTargets="_ChooseAppHost;_GetComHostPaths;_GetIjwHostPaths"
```
Comment (wli3) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:629]: hmm... how that fix the incremental build issue?

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets [hunk 2]
```diff
@@ -506,16 +507,16 @@ Copyright (c) .NET Foundation. All rights reserved.
       <AppHostSourcePath Condition="'$(UseAppHostFromAssetsFile)' == 'true'">@(_NativeRestoredAppHostNETCore)</AppHostSourcePath>
     </PropertyGroup>
     <PropertyGroup Condition="'$(UseAppHostFromAssetsFile)' == 'false' Or '@(_NativeRestoredAppHostNETCore)' != ''">
-      <AppHostIntermediatePath>$([System.IO.Path]::GetFullPath('$(IntermediateOutputPath)$(AssemblyName)$(_NativeExecutableExtension)'))</AppHostIntermediatePath>
+      <AppHostIntermediatePath>$([System.IO.Path]::GetFullPath('$(IntermediateOutputPath)apphost$(_NativeExecutableExtension)'))</AppHostIntermediatePath>
```
Comment (wli3) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:510]: why this change?

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets (hunk 1 above)
Comment (agocke) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:629]: The root problem here is that ChooseAppHost picks either the app host or the single file host, depending on options, and copies one to the intermediate output directory as 'app.exe'. If you switch which app host you are using by changing the options, it still looks like the copy in the intermediate output directory is up to date because it has the same name. 

With this change we now look for app host.exe or singlefilehost.exe, so if you switch the property you'll look for a different file, and the up to date check will work properly. This relies on publish cleaning the final output directory, but that seems to work fine.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets (hunk 2 above)
Comment (wli3) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:510]: this is the path that apphost is decided, so we can just write apphost here?

**Comment by wli3** [comment]
The test failure is caused by the current fire (nuget version mismatch). If the tests passed later. This change should be good to go

**Review by wli3 — APPROVED** [review]

**Comment by wli3** [comment]
nit. It is embarrassing that existing file has extra white spaces. But it is better to separate the white space change with the actual change in different commits.

**Comment by agocke** [comment]
Yeah, the white space changes were accidental, VS Code just does this automatically when I save a file. Unfortunately I don't have a great way to revert.
